consensus/src/payload_manager/quorum_store_payload_manager.rs

```diff
@@ -116,50 +126,129 @@ impl QuorumStorePayloadManager {
-                    .iter()
-                    .map(|proof| proof.info().clone())
-                    .collect::<Vec<_>>(),
-                Payload::InQuorumStoreWithLimit(proof_with_status) => proof_with_status
-                    .proof_with_data
-                    .proofs
-                    .iter()
-                    .map(|proof| proof.info().clone())
-                    .collect::<Vec<_>>(),
+                Payload::InQuorumStore(proof_with_status) => {
+                    for proof in proof_with_status.proofs.iter() {
+                        batches.push(proof.info().clone());
+                    }
+                },
+                Payload::InQuorumStoreWithLimit(proof_with_status) => {
+                    for proof in proof_with_status.proof_with_data.proofs.iter() {
+                        batches.push(proof.info().clone());
+                    }
+                },
                 Payload::QuorumStoreInlineHybrid(inline_batches, proof_with_data, _) => {
-                    inline_batches
-                        .iter()
-                        .map(|(batch_info, _)| batch_info.clone())
-                        .chain(
-                            proof_with_data
-                                .proofs
-                                .iter()
-                                .map(|proof| proof.info().clone()),
-                        )
-                        .collect::<Vec<_>>()
+                    for (batch_info, _) in inline_batches.iter() {
+                        batches.push(batch_info.clone());
+                    }
+                    for proof in proof_with_data.proofs.iter() {
+                        batches.push(proof.info().clone());
+                    }
                 },
                 Payload::OptQuorumStore(opt_quorum_store_payload) => {
-                    opt_quorum_store_payload.into_inner().get_all_batch_infos()
+                    // TODO: how to avoid the clone?
+                    batches = opt_quorum_store_payload
+                        .clone()
+                        .into_inner()
+                        .get_all_batch_infos();
                 },
-            })
+            }
+        }
+        batches
+    }
+
+    fn get_committed_timestamp_and_round(
+        block: &Block,
+        block_window: &OrderedBlockWindow,
+    ) -> (u64, Round) {
+        let block_window_blocks = block_window.blocks().expect("upgrade block window");
+        let blocks: Vec<_> = block_window_blocks
+            .iter()
+            .chain(std::iter::once(block))
             .collect();
+        let oldest_block = blocks.first().expect("at least one block");
+        if block_window.window_start_round() == oldest_block.round() {
+            (oldest_block.timestamp_usecs(), oldest_block.round())
+        } else {
+            let certified_block = oldest_block.quorum_cert().certified_block();
+            (certified_block.timestamp_usecs(), certified_block.round())
+        }
+    }
```

Comment:
Cant this be written more simply like 

```rust
fn get_committed_timestamp_and_round(
    block: &Block,
    block_window: &OrderedBlockWindow,
) -> (u64, Round) {
    let block_window_blocks = block_window.blocks().expect("upgrade block window");

    // Determine the oldest block
    let oldest_block = block_window_blocks.first().unwrap_or(block);

    // Check window start round against the oldest block's round
    if block_window.window_start_round() == oldest_block.round() {
        (oldest_block.timestamp_usecs(), oldest_block.round())
    } else {
        let certified_block = oldest_block.quorum_cert().certified_block();
        (certified_block.timestamp_usecs(), certified_block.round())
    }
}
```

Also should this live under the impl for `OrderedBlockWindow`?

Reply:
LGTM.
Yes, it can be moved to `OrderedBlockWindow` as long as we pass in &Block